chainer/optimizer.py

```diff
@@ -27,153 +30,313 @@ def exponential_decay_noise(xp, shape, dtype, hook, opt):
-    in registration order in advance of the actual parameter update.
+    Args:
+        parent (Hyperparameter): Parent hyperparameter.
+
+    """
+    def __init__(self, parent=None):
+        self._parent = parent
+
+    def __getattr__(self, name):
+        return getattr(self._parent, name)
+
+    def __repr__(self):
+        d = self.get_dict()
+        keys = sorted(d.keys())
+        values_repr = ', '.join('%s=%s' % (k, d[k]) for k in keys)
+        return 'Hyperparameter(%s)' % values_repr
+
+    def get_dict(self):
+        """Converts the hyperparameter into a dictionary.
+
+        Returns:
+            Dictionary containing all entries that can be referred by this
+            hyperparameter object.
+
+        """
+        d = {} if self._parent is None else self._parent.get_dict()
+        for k, v in six.iteritems(self.__dict__):
+            if k != '_parent':
+                d[k] = v
+        return d
+
+    def set_parent(self, parent):
+        """Sets the parent hyperparameter.
+
+        Args:
+            parent (Hyperparameter): The parent hyperparameter to be set.
+
+        """
+        self._parent = parent
+
+
+class UpdateRule(object):
+
+    """Base class of all update rules.
+
+    Update rule is an object that implements how to update one parameter
+    variable using the gradient of a loss function. This class provides the
+    interface and the common features of any update rules.
+
+    An update rule can be set to a :class:`~chainer.Variable` object that
+    represents a parameter array of a model. An :class:`~chainer.Optimizer`
+    instance defines which parameters to update, and the update rule instance
+    of each parameter defines how to update it.
+
+    Hook functions can be set to any update rule instance. The hook function is
+    called just before any updates in the order of registrations.
+
+    An implementation of update rule should override :meth:`update_core` or
+    its device-dependent variants (i.e., :meth:`update_core_cpu` and
+    :meth:`update_core_gpu`).
+
+    The state (e.g. a moving average of the gradient) of the update rule is
+    stored into the state dictionary. An implementation of update rule using
+    state should also override :meth:`init_state` to initialize the state at
+    the first update. The values of the state dictionary are automatically
+    copied to the appropriate device before the update based on the data and
+    grad arrays.
 
     Attributes:
-        target: Target link object. It is set by the :meth:`setup` method.
-        t: Number of update steps. It must be incremented by the
-            :meth:`update` method.
-        epoch: Current epoch. It is incremented by the :meth:`new_epoch`
-            method.
+        enabled (bool): Flag to configure if this update rule is active. If the
+            update rule is not active (i.e., ``enabled = False``), the
+            :meth:`update` method does not update the parameter.
```

Comment:
I often get requests that users want to freeze a part of the model on a specific update without editing the link hierarchy. Using `enabled` flag, we can stop updating parameters without modifying the model definition. I now think it might be better to add a convenient method to `Link` that disables update rules of all parameters under the hierarchy.